rust/xaynet-server/src/state_machine/phases/update.rs

```diff
@@ -223,75 +220,50 @@ impl PhaseState<Update> {
     /// Adds a local seed dictionary to the seed dictionary.
     ///
     /// # Error
-    /// Fails if it contains invalid keys or it is a repetition.
-    fn add_local_seed_dict(
+    /// Fails if the local dict cannot be added due to a Redis error.
+    async fn add_local_seed_dict(
         &mut self,
         pk: &UpdateParticipantPublicKey,
         local_seed_dict: &LocalSeedDict,
     ) -> Result<(), StateMachineError> {
-        if local_seed_dict.keys().len() == self.inner.frozen_sum_dict.keys().len()
```

Comment:
If I'm reading `update_seed_dict` correctly, we're actually removing the following checks

Checking that the seed dictionary sent by the participant has the right length:
```rust
local_seed_dict.keys().len() == self.inner.frozen_sum_dict.keys().len()
```

All of the keys of the seed dictionary sent by the participant have a counterpart in the frozen sum dict:

```rust
local_seed_dict.keys().all(|pk| self.inner.frozen_sum_dict.contains_key(pk))
```

The update participant didn't already send an update:

```rust
self.inner.seed_dict.values().next().map_or(true, |dict| !dict.contains_key(pk))
```

The first two checks can be summarized as "the update dict contains exactly one entry for each of the sum participants". I think this check is important, otherwise we may end up with an invalid seed dict. How doable is this with Redis?

Reply:
yep, you are right. These checks are what my comment in #416 referred to: 
>
        // This means, that Redis does not abort a transaction if for example the
        // command "hset_nx" returns `0` (value already exists).
        //
        // As a result, this method is successful even though one or all of
        // the following points apply:
        // - `sum_pk` is not in the `sum_dict`
        // - `update_pk` already exists (however, in this case the seed is not updated)
        // - the `LocalSeedDict` has a different length/update_pks than the `sum_dict`.
        //
        // We can create our own transactions via lua scripts. In a script we can
        // perform all the checks first and only write the data if the checks were successful.
        // However, in our case, the checks can be quite expensive.
        // Therefore we should agree on which checks are necessary.

And one solution (not sure if this is the only one) is to use lua scripts. I tried to implement these checks in #409. If you want to take a look, [here is the line where it starts](https://github.com/xaynetwork/xaynet/pull/409/files#diff-a2ad71b28c4ae09ddfa8e791350b705aR10)

Reply:
Ok so it's definitely possible to re-add these checks. They shouldn't be too costly if running in a lua script. Do you have short term plans for adding such a script?

Reply:
The lua scripts are task I would like to tackle next. I haven't added them yet because our sprint is so short and I didn't know how long the integration would take.